lib/block_view.go

```diff
@@ -1192,12 +1199,45 @@ func (bav *UtxoView) _verifySignature(txn *MsgDeSoTxn, blockHeight uint32) (_der
 	return nil, RuleErrorInvalidTransactionSignature
 }
 
-func IsDerivedSignature(txn *MsgDeSoTxn) (_derivedPkBytes []byte, _isDerived bool) {
-	if txn.ExtraData == nil {
-		return nil, false
+// IsDerivedSignature checks if a transaction was signed using a derived key. If so, it will recover the derived key used
+// to sign the transaction. There are two possible ways to serialize transaction's ECDSA signature for a derived key.
+// Either to use the DER encoding and place the derived public key in transaction's ExtraData, or to use DeSo-DER signature
+// encoding and pass a special recovery ID into the signature's bytes. However, both encodings can't be used at the same time.
+func IsDerivedSignature(txn *MsgDeSoTxn) (_derivedPkBytes []byte, _isDerived bool, _err error) {
+	// If transaction contains ExtraData, then check if the DerivedPublicKey was passed along.
+	if txn.ExtraData != nil {
+		derivedPkBytes, isDerived := txn.ExtraData[DerivedPublicKey]
+		if isDerived {
+			// Make sure both encodings aren't used concurrently.
+			if txn.Signature.IsRecoverable {
+				return nil, false, errors.Wrapf(RuleErrorDerivedKeyHasBothExtraDataAndRecoveryId,
+					"IsDerivedSignature: transaction signed with a derived key can either store public key in "+
+						"ExtraData or use the DeSo-DER recoverable signature encoding")
+			}
+			return derivedPkBytes, isDerived, nil
+		}
```

Comment:
I would prefer not to nest If statements this deeply, it increase cognitive complexiity.
```suggestion
		// Make sure both encodings aren't used concurrently.
		if isDerived  && txn.Signature.IsRecoverable {
			return nil, false, errors.Wrapf(RuleErrorDerivedKeyHasBothExtraDataAndRecoveryId,
				"IsDerivedSignature: transaction signed with a derived key can either store public key in "+
					"ExtraData or use the DeSo-DER recoverable signature encoding")
		}
		if isDerived {
			return derivedPkBytes, isDerived, nil
		}
```